### ISSUE TITLE
watchdog killing process on boot when running with very large RLIMIT_NOFILE and auto global.maxconn

### ISSUE THREAD

**Issue by apoikos** [haproxy/haproxy#2621]
### Detailed Description of the Problem

This was [reported](https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=1074374) in the Debian Bug Tracker. In short, if a HAProxy process is started with a very large number of RLIMIT_NOFILE.rlim_cur and without explicitly specifying a modest global.maxconn/global.maxsock value, then the process will tend to allocate huge amounts of memory and run in an OOM condition.

### Expected Behavior

~~This is arguably hard to come by as it needs both, a very high fs.nr_open value and a high RLIMIT_NOFILE soft limit.~~ This is triggered by a recent change in systemd, see below.

However I think HAProxy should place a reasonable upper limit (say 200k) to automatically computed global.maxconn/global.maxsock to avoid allocating huge amounts of memory.

### Steps to Reproduce the Behavior

1. `sysctl -w fs.nr_open=1000000000`
2. `ulimit -n 1000000000`
3. `haproxy -V -f /dev/null`

At step 3, HAProxy outputs:
```
Note: setting global.maxconn to 499999983.
Available polling systems :
      epoll : pref=300,  test result OK
       poll : pref=200,  test result OK
     select : pref=150,  test result FAILED
Total: 3 (2 usable), will use epoll.

Available filters :
	[BWLIM] bwlim-in
	[BWLIM] bwlim-out
	[CACHE] cache
	[COMP] compression
	[FCGI] fcgi-app
	[  OT] opentracing
	[SPOE] spoe
	[TRACE] trace
```
And then consumes all system CPU and memory while populating `fdtab` in `init_pollers`. At this point the process's VSZ is around 160GB on my (amd64) system.

### Do you have any idea what may have caused this?
Recently systemd chaged to [bumping the hard fd limit to the kernel max on boot](https://lists.debian.org/debian-devel/2024/06/msg00041.html), which causes unexpected results in HAProxy when falling back to `compute_ideal_maxconn()`.

`compute_ideal_maxconn()` tries to guess how many connections fit in RLIMIT_NOFILE.rlim_cur, but does not take memory into account. As a result when given an absurdly high value for RLIMIT_NOFILE, it will come up with an absurdly high value for maxconn, which will also influence maxsock and the various associated memory allocations.

### Do you have an idea how to solve the issue?

There is already an upper fd bound implemented using `fd-hard-limit`, but `fd-hard-limit` is unset by default. I believe that `fd-hard-limit` should have a "reasonable" default value (say 200k) and print a clear warning when it kicks in  (i.e. when auto-sizing would lead to greater values). This would keep memory usage in check regardless of system limits, and would also be user-configurable in the odd event someone needed a very high FD count indeed.

### What is your configuration?

```haproxy
Any config will do, as long as it does *not* specify `global.maxconn` / `global.maxsock`.
```

### Output of `haproxy -vv`

```plain
HAProxy version 2.9.9-2 2024/06/28 - https://haproxy.org/
Status: stable branch - will stop receiving fixes around Q1 2025.
Known bugs: http://www.haproxy.org/bugs/bugs-2.9.9.html
Running on: Linux 6.6.15-amd64 #1 SMP PREEMPT_DYNAMIC Debian 6.6.15-2 (2024-02-04) x86_64
Build options :
  TARGET  = linux-glibc
  CPU     = generic
  CC      = cc
  CFLAGS  = -O2 -g -O2 -Werror=implicit-function-declaration -fstack-protector-strong -fstack-clash-protection -Wformat -Werror=format-security -fcf-protection -Wdate-time -D_FORTIFY_SOURCE=2 -Wall -Wextra -Wundef -Wdeclaration-after-statement -Wfatal-errors -Wtype-limits -Wshift-negative-value -Wshift-overflow=2 -Wduplicated-cond -Wnull-dereference -fwrapv -Wno-address-of-packed-member -Wno-unused-label -Wno-sign-compare -Wno-unused-parameter -Wno-clobbered -Wno-missing-field-initializers -Wno-cast-function-type -Wno-string-plus-int -Wno-atomic-alignment
  OPTIONS = USE_OPENSSL=1 USE_LUA=1 USE_SLZ=1 USE_SYSTEMD=1 USE_OT=1 USE_QUIC=1 USE_PROMEX=1 USE_PCRE2=1 USE_PCRE2_JIT=1 USE_QUIC_OPENSSL_COMPAT=1
  DEBUG   = -DDEBUG_STRICT -DDEBUG_MEMORY_POOLS

Feature list : -51DEGREES +ACCEPT4 +BACKTRACE -CLOSEFROM +CPU_AFFINITY +CRYPT_H -DEVICEATLAS +DL -ENGINE +EPOLL -EVPORTS +GETADDRINFO -KQUEUE -LIBATOMIC +LIBCRYPT +LINUX_CAP +LINUX_SPLICE +LINUX_TPROXY +LUA +MATH -MEMORY_PROFILING +NETFILTER +NS -OBSOLETE_LINKER +OPENSSL -OPENSSL_AWSLC -OPENSSL_WOLFSSL +OT -PCRE +PCRE2 +PCRE2_JIT -PCRE_JIT +POLL +PRCTL -PROCCTL +PROMEX -PTHREAD_EMULATION +QUIC +QUIC_OPENSSL_COMPAT +RT +SHM_OPEN +SLZ +SSL -STATIC_PCRE -STATIC_PCRE2 +SYSTEMD +TFO +THREAD +THREAD_DUMP +TPROXY -WURFL -ZLIB

Default settings :
  bufsize = 16384, maxrewrite = 1024, maxpollevents = 200

Built with multi-threading support (MAX_TGROUPS=16, MAX_THREADS=256, default=8).
Built with OpenSSL version : OpenSSL 3.2.2 4 Jun 2024
Running on OpenSSL version : OpenSSL 3.2.2-dev 
OpenSSL library supports TLS extensions : yes
OpenSSL library supports SNI : yes
OpenSSL library supports : TLSv1.0 TLSv1.1 TLSv1.2 TLSv1.3
OpenSSL providers loaded : default
Built with Lua version : Lua 5.4.6
Built with the Prometheus exporter as a service
Built with network namespace support.
Built with OpenTracing support.
Built with libslz for stateless compression.
Compression algorithms supported : identity("identity"), deflate("deflate"), raw-deflate("deflate"), gzip("gzip")
Built with transparent proxy support using: IP_TRANSPARENT IPV6_TRANSPARENT IP_FREEBIND
Built with PCRE2 version : 10.42 2022-12-11
PCRE2 library supports JIT : yes
Encrypted password support via crypt(3): yes
Built with gcc compiler version 13.2.0

Available polling systems :
      epoll : pref=300,  test result OK
       poll : pref=200,  test result OK
     select : pref=150,  test result OK
Total: 3 (3 usable), will use epoll.

Available multiplexer protocols :
(protocols marked as <default> cannot be specified using 'proto' keyword)
       quic : mode=HTTP  side=FE     mux=QUIC  flags=HTX|NO_UPG|FRAMED
         h2 : mode=HTTP  side=FE|BE  mux=H2    flags=HTX|HOL_RISK|NO_UPG
       fcgi : mode=HTTP  side=BE     mux=FCGI  flags=HTX|HOL_RISK|NO_UPG
  <default> : mode=HTTP  side=FE|BE  mux=H1    flags=HTX
         h1 : mode=HTTP  side=FE|BE  mux=H1    flags=HTX|NO_UPG
  <default> : mode=TCP   side=FE|BE  mux=PASS  flags=
       none : mode=TCP   side=FE|BE  mux=PASS  flags=NO_UPG

Available services : prometheus-exporter
Available filters :
	[BWLIM] bwlim-in
	[BWLIM] bwlim-out
	[CACHE] cache
	[COMP] compression
	[FCGI] fcgi-app
	[  OT] opentracing
	[SPOE] spoe
	[TRACE] trace


```
```


### Last Outputs and Backtraces

_No response_

### Additional Information

_No response_

**Comment by apoikos** [comment]
The missing bit from the puzzle is that systemd/256~rc3-3 now [sets the file descriptor hard limit to kernel max on boot](https://lists.debian.org/debian-devel/2024/06/msg00041.html), so I would expect this issue to suddenly pop up more often.

**Comment by einval22** [comment]
Hi Apollon !

Thanks a lot for reporting this and for the steps to reproduce !

I've reproduced it as well and we've had a quite similar behavior during the last bench test sessions.

Would you please,  try with the new released version 3.0 and launch haproxy with `-m` option, [here](http://docs.haproxy.org/dev/management.html#3) are more details about this option.

Very high VSZ usage in previous versions under the high load is explained and fixed in this commit [here](https://github.com/haproxy/haproxy/commit/865db6307fb03237e29cd29081248c816f2e22d7).

**Comment by einval22** [comment]
> There is already an upper fd bound implemented using `fd-hard-limit`, but `fd-hard-limit` is unset by default. I believe that `fd-hard-limit` should have a "reasonable" default value (say 200k) and print a clear warning when it kicks in (i.e. when auto-sizing would lead to greater values). This would keep memory usage in check regardless of system limits, and would also be user-configurable in the odd event someone needed a very high FD count indeed.

Extremely huge hard limit of the RLIMIT_NOFILE is one of the cases, why the `-m` option was implemented, in order that haproxy process has never been killed by OOM. 

`Note: setting global.maxconn to 499999983.`

Above is an stdout message from this part of main() code:

```
	ideal_maxconn = compute_ideal_maxconn();

	if (!global.rlimit_memmax) {
		if (global.maxconn == 0) {
			global.maxconn = ideal_maxconn;
			if (global.mode & (MODE_VERBOSE|MODE_DEBUG))
				fprintf(stderr, "Note: setting global.maxconn to %d.\n", global.maxconn);
		}
	}
```

In your case `global.rlimit_memmax = 0`, as it is set only, when we start with -m ([here](https://github.com/haproxy/haproxy/blob/master/src/haproxy.c#L2025)) and [here](https://github.com/haproxy/haproxy/blob/master/src/haproxy.c#L2335).

If you will start with `-m` and `-f /dev/null` (as you did for reproducing),  haproxy will take in account the provided via cmd line memory limit, while calculating the maxconn [here](https://github.com/haproxy/haproxy/blob/master/src/haproxy.c#L2582).

Hope this will help,

Kind regards,

**Comment by einval22** [comment]
Hi Apollon !

I've just saw the bug report in the Debian tracker.

We are looking how to fix this issue better for packages, because ` -m <some reasonable value>` it is basically the same as` fd-hard-limit=<some reasonable value>` in the haproxy unit-file. But this is not the best solution.

Thanks a lot again for reporting this !

**Comment by einval22** [comment]
Hi Apollon !

We've just merged a fix for this issue [here](https://github.com/haproxy/haproxy/commit/41275a691839df5f8dc7cb9faa4e259fbb755d34).

So, you could just compile haproxy for package with build option, which you usually use, ` fd-hard-limit ` will be 1048576 by default.

If you will have some particular platform, where you estimate that you need to hard code an another default, you can redefine it at a build time via:

` make -j 8 TARGET=linux-glibc DEBUG=-DDEFAULT_MAXFD=50000`

Please, feel free to come back if you have some troubles or any suggestions how we could improve the Makefile and other stuff, that you are using for packaging haproxy.

Kind regards,

**Comment by apoikos** [comment]
Thanks for the fast response and the fix @vkssv! I believe that this is a more elegant solution than relying on `-m` , as it makes HAProxy work out of the box with minimal configuration without sacrificing any functionality (I guess 1M fds is still large enough for 99.9% of the installations out there).

**Comment by lukastribus** [comment]
Although it doesn't really matter, `256~rc3-3` is not a systemd version/tag, but a debian package upload of systemd [1]:
  
    systemd (256~rc3-3) unstable; urgency=medium
    [...]
      * Restore open files limit bump on boot. Broken packages ought to have
        been fixed by now. (Closes: #1029152)
    [...]
     -- Luca Boccassi <bluca@debian.org>  Tue, 28 May 2024 00:07:57 +0100


The triggering change comes from debian as per bug #1029152 [2] changed in [3], it's not a change from upstream systemd (the option in systemd is there for some time already).

The maximum would actually be 2B, but systemd set's the limit higher (`2147483640`) than the kernel allows (`2147483584`) [4], so it's rejected which leads systemd to then retry with half the value (`1073741816`) [5]. This is how we get the 1B number we have now.

So at some point we will get a 2B limit instead, if systemd developers copy over the kernel logic to determine the maximum number, as it appears they are intending to do in [4].

None of those details really matter for us ... this is just for context.

(and a note to myself: this is the english "billion" - 10^9 )


[1] https://metadata.ftp-master.debian.org/changelogs//main/s/systemd/systemd_256.2-1_changelog
[2] https://bugs-devel.debian.org/cgi-bin/bugreport.cgi?bug=1029152
[3] https://salsa.debian.org/systemd-team/systemd/-/commit/99066f931bb49b43e7282fc1fe8488373bfb81e5
[4] https://github.com/systemd/systemd/pull/32740
[5] https://github.com/systemd/systemd/blob/b209fff3a51b8e57ea5970575bf8d4db4790b620/src/core/main.c#L1293
